File: go/internal/tex/image.go
package tex

import (
	"errors"
	"fmt"
	"io"
	"strconv"

	"github.com/1m188/repkg-go/internal/binutil"
	"github.com/1m188/repkg-go/pkg/mipmap"
)

// 安全限制常量
const maxMipmapCount = 32
const maxMipmapByteCount = 250_000_000 // 250 MB
const maxImageCount = 100

// Mipmap 表示纹理的一个 mipmap 级别。
// 对应 C# TexMipmap。
type Mipmap struct {
	Bytes                  []byte        // 像素数据（可能压缩）
	Width                  int32         // 宽度
	Height                 int32         // 高度
	DecompressedBytesCount int32         // LZ4 解压后的预期字节数
	IsLZ4Compressed        bool          // 是否 LZ4 压缩
	Format                 mipmap.Format // 像素格式
}

// Image 表示 TEX 中的一张图片（含 mipmap 链）。
// 对应 C# TexImage。
type Image struct {
	Mipmaps []*Mipmap // Mipmaps mipmap 级别链。
}

// FirstMipmap 返回第一个 mipmap 级别，如果没有 mipmap 则返回 nil。
func (img *Image) FirstMipmap() *Mipmap {
	if len(img.Mipmaps) == 0 {
		return nil
	}
	return img.Mipmaps[0]
}

// ==================== Mipmap 读取 ====================

// readMipmapV1 读取版本 1 的 mipmap（无 LZ4 压缩字段）。
// skipBytes 为 true 时跳过字节数据（用于 ReadMipmapBytes=false 模式）。
func readMipmapV1(r byteReader, skipBytes bool) (*Mipmap, error) {
	w, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取mipmap宽度失败: %w", err)
	}
	h, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取mipmap高度失败: %w", err)
	}
	bytes, err := readMipmapBytes(r, skipBytes)
	if err != nil {
		return nil, err
	}
	return &Mipmap{Width: w, Height: h, Bytes: bytes}, nil
}

// readMipmapV2And3 读取版本 2/3 的 mipmap（含 LZ4 压缩字段）。
// skipBytes 为 true 时跳过字节数据（用于 ReadMipmapBytes=false 模式）。
func readMipmapV2And3(r byteReader, skipBytes bool) (*Mipmap, error) {
	w, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取mipmap宽度失败: %w", err)
	}
	h, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取mipmap高度失败: %w", err)
	}
	isLZ4, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取LZ4标志失败: %w", err)
	}
	decompCount, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取解压后字节数失败: %w", err)
	}
	bytes, err := readMipmapBytes(r, skipBytes)
	if err != nil {
		return nil, err
	}
	return &Mipmap{
		Width:                  w,
		Height:                 h,
		IsLZ4Compressed:        isLZ4 == 1,
		DecompressedBytesCount: decompCount,
		Bytes:                  bytes,
	}, nil
}

// readMipmapV4 读取版本 4 的 mipmap（含 conditionJson）。
func readMipmapV4(r ioRuneReader, skipBytes bool) (*Mipmap, error) {
	param1, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取param1失败: %w", err)
	}
	if param1 != 1 {
		return nil, &UnknownMagicError{Source: SourceReadMipmapV4, Property: "param1", Magic: strconv.Itoa(int(param1))}
	}
	param2, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取param2失败: %w", err)
	}
	if param2 != 2 {
		return nil, &UnknownMagicError{Source: SourceReadMipmapV4, Property: "param2", Magic: strconv.Itoa(int(param2))}
	}

	// 读取 conditionJson（null 结尾字符串）
	_, err = binutil.ReadNString(r, -1)
	if err != nil {
		return nil, fmt.Errorf("读取 conditionJson 失败: %w", err)
	}

	param3, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取param3失败: %w", err)
	}
	if param3 != 1 {
		return nil, &UnknownMagicError{Source: SourceReadMipmapV4, Property: "param3", Magic: strconv.Itoa(int(param3))}
	}

	return readMipmapV2And3(r, skipBytes)
}

// readMipmapBytes 读取 mipmap 的字节数据（int32 长度前缀 + 数据体）。
// 当 skipBytes 为 true 时跳过字节而不返回数据（此情况返回 nil）。
func readMipmapBytes(r byteReader, skipBytes bool) ([]byte, error) { //nolint:revive // skipBytes 是性能优化标志，由调用方决定
	byteCount, err := binutil.ReadInt32(r)
	if err != nil {
		return nil, fmt.Errorf("读取 mipmap 字节数失败: %w", err)
	}

	if byteCount < 0 {
		return nil, &UnsafeTexError{Reason: fmt.Sprintf("mipmap 字节数无效: %d", byteCount)}
	}
	if byteCount > maxMipmapByteCount {
		return nil, &UnsafeTexError{Reason: fmt.Sprintf("mipmap 字节数超出限制: %d / %d", byteCount, maxMipmapByteCount)}
	}

	if skipBytes {
		// 跳过字节但不分配内存
		skipBuf := make([]byte, 4096)
		remaining := int(byteCount)
		for remaining > 0 {
			chunk := skipBuf
			if remaining < len(chunk) {
				chunk = skipBuf[:remaining]
			}
			n, readErr := r.Read(chunk)
			if readErr != nil {
				return nil, fmt.Errorf("跳过 mipmap 数据失败: %w", readErr)
			}
			if n == 0 {
				return nil, &UnsafeTexError{Reason: "reader 返回 0 字节但无错误，无法继续跳过 mipmap 数据"}
			}
			remaining -= n
		}
		return nil, nil
	}

	buf := make([]byte, byteCount)
	_, err = io.ReadFull(r, buf)
	if err != nil {
		return nil, fmt.Errorf("读取 mipmap 数据失败: %w", err)
	}
	return buf, nil
}

// ==================== Mipmap 写入 ====================

// writeMipmapV1 写入版本 1 的 mipmap。V1 容器不支持 LZ4 压缩数据。
func (m *Mipmap) writeV1(w byteWriter) error {
	if m.IsLZ4Compressed {
		return errors.New("V1 容器不支持 LZ4 压缩的 mipmap")
	}
	err := binutil.WriteInt32(w, m.Width)
	if err != nil {
		return fmt.Errorf("写入mipmap宽度失败: %w", err)
	}
	err = binutil.WriteInt32(w, m.Height)
	if err != nil {
		return fmt.Errorf("写入mipmap高度失败: %w", err)
	}
	return writeMipmapBytes(w, m.Bytes)
}

// writeMipmapV2 写入版本 2+ 的 mipmap（含 LZ4 字段）。
func (m *Mipmap) writeV2(w byteWriter) error {
	err := binutil.WriteInt32(w, m.Width)
	if err != nil {
		return fmt.Errorf("写入mipmap宽度失败: %w", err)
	}
	err = binutil.WriteInt32(w, m.Height)
	if err != nil {
		return fmt.Errorf("写入mipmap高度失败: %w", err)
	}

	lz4Flag := int32(0)
	if m.IsLZ4Compressed {
		lz4Flag = 1
	}
	err = binutil.WriteInt32(w, lz4Flag)
	if err != nil {
		return fmt.Errorf("写入LZ4标志失败: %w", err)
	}
	err = binutil.WriteInt32(w, m.DecompressedBytesCount)
	if err != nil {
		return fmt.Errorf("写入解压后字节数失败: %w", err)
	}
	return writeMipmapBytes(w, m.Bytes)
}

// writeMipmapBytes 写入 mipmap 的字节数据。
func writeMipmapBytes(w byteWriter, bytes []byte) error {
	err := binutil.WriteInt32(w, int32(len(bytes))) //nolint:gosec // bytes 长度受限于上游数据大小
	if err != nil {
		return fmt.Errorf("写入mipmap字节数失败: %w", err)
	}
	_, err = w.Write(bytes)
	if err != nil {
		return fmt.Errorf("写入mipmap数据失败: %w", err)
	}
	return nil
}

File: rs/src/tex/image.rs
// Image 和 Mipmap 数据模型，以及 V1/V2/V3/V4 格式的 mipmap 读写。
// 与 C# 原版 RePKG.Core/Texture/TexImage.cs + TexMipmap.cs 和 RePKG.Application/Texture/TexImageReader.cs 对应。

use std::io::{self, Read, Write};

use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};

use crate::binutil;
use crate::error::UnsafeTexError;
use crate::format::MipmapFormat;

/// 单个 mipmap 级别数据。
#[derive(Debug, Clone)]
pub struct Mipmap {
    /// 像素或压缩数据。
    pub bytes: Vec<u8>,
    /// 宽度。
    pub width: i32,
    /// 高度。
    pub height: i32,
    /// LZ4 解压后的预期字节数。
    pub decompressed_bytes_count: i32,
    /// 是否经过 LZ4 压缩。
    pub is_lz4_compressed: bool,
    /// mipmap 像素格式。
    pub format: MipmapFormat,
    /// V4 容器中的 conditionJson（保留回写）。
    pub condition_json: Option<String>,
}

/// 单张图像，包含一组 mipmap 链。
#[derive(Debug, Clone)]
pub struct Image {
    /// mipmap 链（[0] 为最高分辨率）。
    pub mipmaps: Vec<Mipmap>,
}

/// mipmap 安全常量。
const MAX_MIPMAP_COUNT: usize = 32;
const MAX_MIPMAP_BYTE_COUNT: usize = 250_000_000;

/// 读取 V1 格式的 mipmap（无 LZ4 支持）。
pub fn read_mipmap_v1<R: Read>(reader: &mut R, format: MipmapFormat) -> io::Result<Mipmap> {
    let width = reader.read_i32::<LittleEndian>()?;
    let height = reader.read_i32::<LittleEndian>()?;
    let byte_count = reader.read_i32::<LittleEndian>()?;

    if width < 0 || height < 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!("mipmap 尺寸无效：{}x{}", width, height),
            }
            .to_string(),
        ));
    }

    if byte_count < 0 || byte_count as usize > MAX_MIPMAP_BYTE_COUNT {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!("mipmap 字节数超限：{}", byte_count),
            }
            .to_string(),
        ));
    }

    let mut bytes = vec![0u8; byte_count as usize];
    reader.read_exact(&mut bytes)?;

    Ok(Mipmap {
        bytes,
        width,
        height,
        decompressed_bytes_count: 0,
        is_lz4_compressed: false,
        format,
        condition_json: None,
    })
}

/// 读取 V2/V3 格式的 mipmap（含 LZ4 支持）。
pub fn read_mipmap_v2_v3<R: Read>(reader: &mut R, format: MipmapFormat) -> io::Result<Mipmap> {
    let width = reader.read_i32::<LittleEndian>()?;
    let height = reader.read_i32::<LittleEndian>()?;
    let is_lz4_val = reader.read_i32::<LittleEndian>()?;
    let is_lz4_compressed = is_lz4_val == 1;
    let decompressed_bytes_count = reader.read_i32::<LittleEndian>()?;
    let byte_count = reader.read_i32::<LittleEndian>()?;

    if width < 0 || height < 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!("mipmap 尺寸无效：{}x{}", width, height),
            }
            .to_string(),
        ));
    }
    if decompressed_bytes_count < 0 || decompressed_bytes_count as usize > MAX_MIPMAP_BYTE_COUNT {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!(
                    "mipmap 解压后字节数超限：{}（最大 {}）",
                    decompressed_bytes_count, MAX_MIPMAP_BYTE_COUNT
                ),
            }
            .to_string(),
        ));
    }

    if byte_count < 0 || byte_count as usize > MAX_MIPMAP_BYTE_COUNT {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!("mipmap 字节数超限：{}", byte_count),
            }
            .to_string(),
        ));
    }

    let mut bytes = vec![0u8; byte_count as usize];
    reader.read_exact(&mut bytes)?;

    Ok(Mipmap {
        bytes,
        width,
        height,
        decompressed_bytes_count,
        is_lz4_compressed,
        format,
        condition_json: None,
    })
}

/// 读取 V4 格式的 mipmap（含视频前置字段）。
pub fn read_mipmap_v4<R: Read>(reader: &mut R, format: MipmapFormat) -> io::Result<Mipmap> {
    let param1 = reader.read_i32::<LittleEndian>()?;
    let param2 = reader.read_i32::<LittleEndian>()?;
    let condition_json = binutil::read_n_string(reader, 256)?;
    let param3 = reader.read_i32::<LittleEndian>()?;

    if param1 != 1 || param2 != 2 || param3 != 1 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!(
                    "V4 mipmap 参数错误：param1={}, param2={}, param3={}",
                    param1, param2, param3
                ),
            }
            .to_string(),
        ));
    }

    let mut mipmap = read_mipmap_v2_v3(reader, format)?;
    mipmap.condition_json = if condition_json.is_empty() {
        None
    } else {
        Some(condition_json)
    };
    Ok(mipmap)
}

/// 写入 V1 格式的 mipmap（仅适用于未压缩数据）。
pub fn write_mipmap_v1<W: Write>(writer: &mut W, mipmap: &Mipmap) -> io::Result<()> {
    writer.write_i32::<LittleEndian>(mipmap.width)?;
    writer.write_i32::<LittleEndian>(mipmap.height)?;
    writer.write_i32::<LittleEndian>(mipmap.bytes.len() as i32)?;
    writer.write_all(&mipmap.bytes)?;
    Ok(())
}

/// 写入 V2/V3 格式的 mipmap（含 LZ4 信息）。
pub fn write_mipmap_v2_v3<W: Write>(writer: &mut W, mipmap: &Mipmap) -> io::Result<()> {
    writer.write_i32::<LittleEndian>(mipmap.width)?;
    writer.write_i32::<LittleEndian>(mipmap.height)?;
    writer.write_i32::<LittleEndian>(if mipmap.is_lz4_compressed { 1 } else { 0 })?;
    writer.write_i32::<LittleEndian>(mipmap.decompressed_bytes_count)?;
    writer.write_i32::<LittleEndian>(mipmap.bytes.len() as i32)?;
    writer.write_all(&mipmap.bytes)?;
    Ok(())
}

/// 写入 V4 格式的 mipmap（含视频前置字段）。
pub fn write_mipmap_v4<W: Write>(writer: &mut W, mipmap: &Mipmap) -> io::Result<()> {
    writer.write_i32::<LittleEndian>(1)?; // param1
    writer.write_i32::<LittleEndian>(2)?; // param2
    let cj = mipmap.condition_json.as_deref().unwrap_or("");
    binutil::write_n_string(writer, cj)?; // conditionJson
    writer.write_i32::<LittleEndian>(1)?; // param3
    write_mipmap_v2_v3(writer, mipmap)
}

/// 读取单张 Image（mipmapCount + mipmap 链）。
pub fn read_image<R: Read>(
    reader: &mut R,
    version: super::header::ImageContainerVersion,
    format: MipmapFormat,
) -> io::Result<Image> {
    let mipmap_count = reader.read_i32::<LittleEndian>()?;
    if mipmap_count < 0 || mipmap_count as usize > MAX_MIPMAP_COUNT {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            UnsafeTexError {
                message: format!("mipmap 数量超限：{}", mipmap_count),
            }
            .to_string(),
        ));
    }

    let mut mipmaps = Vec::with_capacity(mipmap_count as usize);
    for _ in 0..mipmap_count {
        let mipmap = match version {
            super::header::ImageContainerVersion::Version1 => read_mipmap_v1(reader, format)?,
            super::header::ImageContainerVersion::Version4 => read_mipmap_v4(reader, format)?,
            _ => read_mipmap_v2_v3(reader, format)?,
        };
        mipmaps.push(mipmap);
    }

    Ok(Image { mipmaps })
}

/// 写入单张 Image。
pub fn write_image<W: Write>(
    writer: &mut W,
    image: &Image,
    version: super::header::ImageContainerVersion,
) -> io::Result<()> {
    writer.write_i32::<LittleEndian>(image.mipmaps.len() as i32)?;

    for mipmap in &image.mipmaps {
        match version {
            super::header::ImageContainerVersion::Version1 => write_mipmap_v1(writer, mipmap)?,
            super::header::ImageContainerVersion::Version4 => write_mipmap_v4(writer, mipmap)?,
            _ => write_mipmap_v2_v3(writer, mipmap)?,
        };
    }

    Ok(())
}
